src/pip/_internal/vcs/helix_core.py

```diff
@@ -0,0 +1,286 @@
+from __future__ import absolute_import
+
+import logging
+import os
+import random
+import re
+import socket
+import string
+
+from pip._vendor.six.moves.urllib import parse as urllib_parse
+from pip._internal.vcs import VersionControl, vcs
+
+logger = logging.getLogger(__name__)
+
+client_spec_template = """
+Owner:   {user}
+Host:    {host}
+Client:  {name}
+Root:    {client_path}
+View:    /{server_path}/... //{name}/...
+Options: allwrite rmdir
+"""
+
+
+def split_url(url, environ=os.environ):
+    """
+    Parses a URL and returns a 5-tuple of
+    ``(port, depot_path, revision, username, password)``. Where fields are not
+    present in the given URL, information from Helix Core environment variables
+    (``P4PORT``, ``P4USER``, ``P4PASSWD``) is used.
+
+    URL examples::
+
+        ssl://me:p4ssw0rd@perforce1/depot/myproject
+        p4://perforce1:1667/depot/myproject
+        p4:///depot/myproject
+    """
+
+    if url.startswith("p4+"):
+        url = url[3:]
+    url = urllib_parse.urlsplit(url)
+
+    p4path = url.path
+    if '@' in p4path:
+        p4path, p4rev = p4path.split("@")
+    else:
+        p4rev = None
+
+    p4user = (
+        url.username or
+        environ.get("P4USER") or
+        environ.get("USER") or
+        environ.get("USERNAME") or
+        '')
+    p4passwd = (
+        url.password or
+        environ.get("P4PASSWD") or
+        '')
+
+    p4port = ["tcp", "perforce", "1666"]
+    p4port_old = environ.get("P4PORT")
+    if p4port_old:
+        for i, v in enumerate(reversed(p4port_old.split(":"))):
+            if v:
+                p4port[-i - 1] = v
+    if url.scheme != "p4":
+        p4port[0] = url.scheme
+    if url.hostname:
+        p4port[1] = url.hostname
+    if url.port:
+        p4port[2] = str(url.port)
```

Comment:
It seems like it would be more explicit / easily understandable to define `p4port` once you know all the values, instead of defining a candidate list and then mutating the list.

Reply:
There's three contributing layers, from lowest to highest precedence:

1. The (Perforce-defined) defaults: [protocol, host, port]
2. The contents of P4PORT, which can provide [protocol, host, port], [host, port] or [port]
3. The URL, which can provide any combination of these.

I'll play around with the implementation some more and try to make this more explicit.

Reply:
I just mean make the final line be `p4port = [scheme, hostname, port]` instead of starting out with default values and then changing the values as you go along. It just seems clearer / more explicit that way.